README.md
# AloFramework | Common #

Common components of AloFramework

[![License](https://poser.pugx.org/aloframework/common/license?format=plastic)](https://www.gnu.org/licenses/gpl-3.0.en.html)
[![Latest Stable Version](https://poser.pugx.org/aloframework/common/v/stable?format=plastic)](https://packagist.org/packages/aloframework/common)
[![Total Downloads](https://poser.pugx.org/aloframework/common/downloads?format=plastic)](https://packagist.org/packages/aloframework/common)

Latest release API documentation: [https://aloframework.github.io/common/](https://aloframework.github.io/common/)

|                                                                                         dev-develop                                                                                         |                                                                               Latest release                                                                               |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
|                                [![Dev Build Status](https://travis-ci.org/aloframework/common.svg?branch=develop)](https://travis-ci.org/aloframework/common)                               |                      [![Release Build Status](https://travis-ci.org/aloframework/common.svg?branch=master)](https://travis-ci.org/aloframework/common)                      |
| [![SensioLabsInsight](https://insight.sensiolabs.com/projects/2f980e0c-215e-4d11-8313-4d071b32b4fd/mini.png)](https://insight.sensiolabs.com/projects/2f980e0c-215e-4d11-8313-4d071b32b4fd) |                    [![SensioLabsInsight](https://i.imgur.com/KygqLtf.png)](https://insight.sensiolabs.com/projects/2f980e0c-215e-4d11-8313-4d071b32b4fd)                   |
| [![Coverage Status](https://coveralls.io/repos/aloframework/common/badge.svg?branch=develop&service=github)](https://coveralls.io/github/aloframework/common?branch=develop)                | [![Coverage Status](https://coveralls.io/repos/aloframework/common/badge.svg?branch=master&service=github)](https://coveralls.io/github/aloframework/common?branch=master) |


CHANGELOG.md
# 2.0

Licenced under Apache-2.0 now.

# 1.3.3 #

The following happened to getUniqid():

Default $entropy value set to 10000, a warning is triggered if openssl_random_pseudo_bytes is unable to locate a 
cryptographically strong algorithm.

# 1.3.2 #

ENT_SUBSTITUTE added to Alo::unXss() 

# 1.3.1 #

Fixed a bug in getUniqid() which tried to use a nonexistent constant

# 1.3 #

getUniqid(), asciiRand(), isRegularRequest() added

# 1.2 #

Added unXss(), getFingerprint() and isTraversable()

# 1.1.1 #

Fixed a bug with isAjaxRequest()

# 1.1 #

`Alo::ifundefined()` added


tests/AloTest.php
<?php
    /**
 *    Copyright (c) Arturas Molcanovas <a.molcanovas@gmail.com> 2016.
 *    https://github.com/aloframework/common
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

    namespace AloFramework\Common;

    use PHPUnit_Framework_TestCase;

    class AloTest extends PHPUnit_Framework_TestCase {

        /** @dataProvider pathProvider */
        function testIncludeOnceIfExists($path, $shouldBe) {
            $assert = $shouldBe ? 'assertTrue' : 'assertFalse';

            call_user_func([$this, $assert],
                           Alo::includeOnceIfExists($path),
                           'includeOnceIfExists failed: path ' . $path . ' should be ' . $assert);
        }

        /** @dataProvider pathProvider */
        function testIncludeIfExists($path, $shouldBe) {
            $assert = $shouldBe ? 'assertTrue' : 'assertFalse';

            call_user_func([$this, $assert],
                           Alo::includeIfExists($path),
                           'includeIfExists failed: path ' . $path . ' should be ' . $assert);
        }

        function testIsAjax() {
            $this->assertFalse(Alo::isAjaxRequest());
        }

        function testIsCLI() {
            $this->assertTrue(Alo::isCliRequest());
        }

        function testIfnull() {
            $v1 = 'string';
            $v2 = 0;
            $v3 = false;
            $v4 = null;
            $v5 = '';

            $planB = 'replacementVar';

            $this->assertEquals($v1, Alo::ifnull($v1, $planB, false), '::get() $v1 failed');
            $this->assertEquals($v2, Alo::ifnull($v2, $planB, false), '::get() $v2 failed');
            $this->assertEquals($v3, Alo::ifnull($v3, $planB, false), '::get() $v3 failed');
            $this->assertEquals($planB, Alo::ifnull($v4, $planB, false), '::get() $v4 failed');
            $this->assertEquals($v5, Alo::ifnull($v5, $planB, false), '::get() $v5 failed');
            $this->assertEquals($planB, Alo::ifnull($nope, $planB, false), '::get() $nope failed');

            $this->assertEquals($v1, Alo::ifnull($v1, $planB, true), '::nullget() $v1 failed');
            $this->assertEquals($planB, Alo::ifnull($v2, $planB, true), '::nullget() $v2 failed');
            $this->assertEquals($planB, Alo::ifnull($v3, $planB, true), '::nullget() $v3 failed');
            $this->assertEquals($planB, Alo::ifnull($v4, $planB, true), '::nullget() $v4 failed');
            $this->assertEquals($planB, Alo::ifnull($v5, $planB, true), '::nullget() $v5 failed');
            $this->assertEquals($planB, Alo::ifnull($nope, $planB, true), '::nullget() $nope failed');
        }

        function testGet() {
            $v1 = 'string';
            $v2 = 0;
            $v3 = false;
            $v4 = null;
            $v5 = '';

            $this->assertTrue(Alo::get($v1) !== null, '$v1 failed');
            $this->assertTrue(Alo::get($v2) !== null, '$v2 failed');
            $this->assertTrue(Alo::get($v3) !== null, '$v3 failed');
            $this->assertTrue(Alo::get($v4) === null, '$v4 failed');
            $this->assertTrue(Alo::get($v5) !== null, '$v5 failed');
            $this->assertTrue(Alo::get($nope) === null, '$nope failed');
        }

        function testNullget() {
            $v1 = 'string';
            $v2 = 0;
            $v3 = false;
            $v4 = null;
            $v5 = '';

            $this->assertTrue(Alo::nullget($v1) !== null, '$v1 failed');
            $this->assertTrue(Alo::nullget($v2) === null, '$v2 failed');
            $this->assertTrue(Alo::nullget($v3) === null, '$v3 failed');
            $this->assertTrue(Alo::nullget($v4) === null, '$v4 failed');
            $this->assertTrue(Alo::nullget($v5) === null, '$v5 failed');
            $this->assertTrue(Alo::nullget($nope) === null, '$nope failed');
        }

        function pathProvider() {
            return [[__DIR__ . DIRECTORY_SEPARATOR . 'IncludeTestFile.php',
                     true],
                    [__DIR__, false],
                    ['/nonexistent/' . mt_rand(~PHP_INT_MAX, PHP_INT_MAX), false]];
        }

        function testIfundefined() {
            $const1 = self::generateConstName();
            $const2 = self::generateConstName();
            $planB = 'bar';

            while (defined($const1)) {
                $const1 = self::generateConstName();
            }

            while (defined($const2)) {
                $const2 = self::generateConstName();
            }

            define($const1, 'foo');

            $this->assertEquals(constant($const1), Alo::ifundefined($const1, $planB));
            $this->assertEquals($planB, Alo::ifundefined($const2, $planB));
        }

        /** @dataProvider isTraversableDataProvider */
        function testIsTraversable($input, $expected) {
            $this->assertEquals($expected, Alo::isTraversable($input));
        }

        function testGetFingerprint() {
            $default = Alo::getFingerprint();
            $this->assertEquals(64, strlen($default));
            $this->assertEquals(128, strlen(Alo::getFingerprint('sha512')));
            $this->assertEquals($default, Alo::getFingerprint('sha256'));
        }

        function isTraversableDataProvider() {
            return [['foo', false],
                    [1.1, false],
                    [new \stdClass(), false],
                    [[], true],
                    [new \ArrayObject(), true]];
        }

        function testUnXSS() {
            $strRaw = '"foo"&\'bar\'';
            $strClean = '&quot;foo&quot;&amp;&apos;bar&apos;';
            $arrRaw = ['one' => $strRaw,
                       'two' => $strRaw];
            $arrClean = ['one' => $strClean,
                         'two' => $strClean];

            $this->assertEquals(new \stdClass(), Alo::unXss(new \stdClass()));
            $this->assertEquals('foo', Alo::unXss('foo'));
            $this->assertEquals($strClean, Alo::unXss($strRaw));
            $this->assertEquals($arrClean, Alo::unXss($arrRaw));
            $this->assertEquals(new \ArrayObject($arrClean), Alo::unXss(new \ArrayObject($arrRaw)));
        }

        function testAsciiRand() {
            $alphanum = Alo::asciiRand(1000, Alo::ASCII_ALPHANUM);
            $nonalphanum = Alo::asciiRand(1000, Alo::ASCII_NONALPHANUM);
            $all = Alo::asciiRand(1000, Alo::ASCII_ALL);

            $this->assertEquals(1000, strlen($alphanum));
            $this->assertEquals(1, preg_match('~^[a-z0-9]+$~i', $alphanum));
            $this->assertEquals(0, preg_match('~^[a-z0-9]+$~i', $nonalphanum));
            $this->assertEquals([1, 1], [preg_match('~[a-z0-9]+~i', $all), preg_match('~[^a-z0-9]+~i', $all)]);
        }

        function testIsRegularRequest() {
            $this->assertFalse(Alo::isRegularRequest());
        }

        private static function generateConstName() {
            return hash('sha512',
                        mt_rand(~PHP_INT_MAX, PHP_INT_MAX) . uniqid(mt_rand(~PHP_INT_MAX, PHP_INT_MAX), true));
        }
    }


tests/IncludeTestFile.php
<?php
    /**
 *    Copyright (c) Arturas Molcanovas <a.molcanovas@gmail.com> 2016.
 *    https://github.com/aloframework/common
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

    //This file is empty and is only used for include tests


tests/IncludeTestFileq.php
<?php
    //This file is empty and is only used for include tests


src/Alo.php
<?php
    /**
 *    Copyright (c) Arturas Molcanovas <a.molcanovas@gmail.com> 2016.
 *    https://github.com/aloframework/common
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

    namespace AloFramework\Common;

    use Traversable;

    /**
     * Static common component container
     *
     * @author Art <a.molcanovas@gmail.com>
     * @since  1.3 getUniqid(), asciiRand(), isRegularRequest() added<br/>
     *         1.2 getFingerprint(), isTraversable(), unXss() added<br/>
     *         1.1 ifundefined() added
     */
    abstract class Alo {

        /**
         * Defines the ascii charset subset as "the entire set"
         *
         * @var int
         */
        const ASCII_ALL = 0;

        /**
         * Defines the ascii charset subset as "only alphanumeric"
         *
         * @var int
         */
        const ASCII_ALPHANUM = 1;

        /**
         * Defines the ascii charset subset as "only non-alphanumeric"
         *
         * @var int
         */
        const ASCII_NONALPHANUM = 2;

        /**
         * Array of ASCII alphanumeric characters
         *
         * @var array
         */
        private static $asciiAlphanum = ['a',
                                         'b',
                                         'c',
                                         'd',
                                         'e',
                                         'f',
                                         'g',
                                         'h',
                                         'i',
                                         'j',
                                         'k',
                                         'l',
                                         'm',
                                         'n',
                                         'o',
                                         'p',
                                         'q',
                                         'r',
                                         's',
                                         't',
                                         'u',
                                         'v',
                                         'w',
                                         'x',
                                         'y',
                                         'z',
                                         'A',
                                         'B',
                                         'C',
                                         'D',
                                         'E',
                                         'F',
                                         'G',
                                         'H',
                                         'I',
                                         'J',
                                         'K',
                                         'L',
                                         'M',
                                         'N',
                                         'O',
                                         'P',
                                         'Q',
                                         'R',
                                         'S',
                                         'T',
                                         'U',
                                         'V',
                                         'W',
                                         'X',
                                         'Y',
                                         'Z',
                                         0,
                                         1,
                                         2,
                                         3,
                                         4,
                                         5,
                                         6,
                                         7,
                                         8,
                                         9];
        /**
         * The rest of the ASCII charset
         *
         * @var array
         */
        private static $asciNonAlphanum = [' ',
                                           '!',
                                           '"',
                                           '#',
                                           '$',
                                           '%',
                                           '\'',
                                           '(',
                                           ')',
                                           '*',
                                           '+',
                                           ',',
                                           '.',
                                           '/',
                                           ':',
                                           ';',
                                           '<',
                                           '=',
                                           '>',
                                           '?',
                                           '@',
                                           '[',
                                           '\\',
                                           ']',
                                           '^',
                                           '_',
                                           '`',
                                           '-',
                                           '{',
                                           '|',
                                           '}',
                                           '~'];

        /**
         * Includes a file if it exists
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string $path Path to the file
         *
         * @return bool True if the file is found, false if not
         */
        static function includeIfExists($path) {
            if (self::isIncludable($path)) {
                include $path;

                return true;
            }

            return false;
        }

        /**
         * Generates a string of random ASCII characters
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param int $length The length of the string
         * @param int $subset Which subset to use - see class' ASCII_* constants
         *
         * @return string
         * @since  1.3
         */
        static function asciiRand($length, $subset = self::ASCII_ALL) {
            switch ($subset) {
                case self::ASCII_ALPHANUM:
                    $subset = self::$asciiAlphanum;
                    break;
                case self::ASCII_NONALPHANUM:
                    $subset = self::$asciNonAlphanum;
                    break;
                default:
                    $subset = array_merge(self::$asciiAlphanum, self::$asciNonAlphanum);
            }

            $count = count($subset) - 1;

            $r = '';

            for ($i = 0; $i < $length; $i++) {
                $r .= $subset[mt_rand(0, $count)];
            }

            return $r;
        }

        /**
         * Generates a unique identifier
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string $hash      Hash algorithm
         * @param string $prefix    Prefix for the identifier
         * @param int    $entropy   Number of pseudo bytes used in entropy
         * @param bool   $rawOutput When set to true, outputs raw binary data. false outputs lowercase hexits.
         *
         * @return string
         * @see    https://secure.php.net/manual/en/function.hash.php
         * @see    https://secure.php.net/manual/en/function.openssl-random-pseudo-bytes.php
         * @since  1.3.3 Default $entropy value set to 10000, a warning is triggered if openssl_random_pseudo_bytes is
         * unable to locate a cryptographically strong algorithm.<br/>
         *         1.3
         * @codeCoverageIgnore
         */
        static function getUniqid($hash = 'sha256', $prefix = '', $entropy = 10000, $rawOutput = false) {
            $str = mt_rand(~PHP_INT_MAX, PHP_INT_MAX) . json_encode([$_COOKIE,
                                                                     $_REQUEST,
                                                                     $_FILES,
                                                                     $_ENV,
                                                                     $_GET,
                                                                     $_POST,
                                                                     $_SERVER]) . uniqid($prefix, true) .
                   self::asciiRand($entropy, self::ASCII_ALL);

            if (function_exists('\openssl_random_pseudo_bytes')) {
                $algoStrong = null;
                $str .= \openssl_random_pseudo_bytes($entropy, $algoStrong);

                if ($algoStrong !== true) {
                    trigger_error('Please update your openssl & PHP libraries. openssl_random_pseudo_bytes was unable' .
                                  ' to locate a cryptographically strong algorithm.',
                                  E_USER_WARNING);
                }
            } else {
                trigger_error('The openssl extension is not enabled, therefore the unique ID is not ' .
                              'cryptographically secure.',
                              E_USER_WARNING);
            }

            return hash($hash, $str, $rawOutput);
        }

        /**
         * Checks if the path is includable
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string $path The path
         *
         * @return bool
         */
        private static function isIncludable($path) {
            return file_exists($path) && is_file($path);
        }

        /**
         * include_once() a file if it exists
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string $path Path to the file
         *
         * @return bool True if the file is found, false if not
         */
        static function includeOnceIfExists($path) {
            if (self::isIncludable($path)) {
                include_once $path;

                return true;
            }

            return false;
        }

        /**
         * Checks if we're dealing with a CLI request
         *
         * @author Art <a.molcanovas@gmail.com>
         * @return bool
         */
        static function isCliRequest() {
            return PHP_SAPI == 'cli' || defined('STDIN');
        }

        /**
         * Checks if the request is non-ajax and non-CLI
         *
         * @author Art <a.molcanovas@gmail.com>
         * @return bool
         * @since  1.3
         */
        static function isRegularRequest() {
            return !self::isAjaxRequest() && !self::isCliRequest();
        }

        /**
         * Returns $var if it's set, null if it's not
         *
         * @param mixed $var Reference to the variable
         *
         * @return mixed|null $var if it's set, null if it's not
         */
        static function get(&$var) {
            return isset($var) ? $var : null;
        }

        /**
         * Returns $var if it's set and evaluates as true (a non-empty string, non-0 int etc), null otherwise
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param mixed $var Reference to the variable
         *
         * @return mixed|null The var or null
         */
        static function nullget(&$var) {
            return self::get($var) ? $var : null;
        }

        /**
         * Returns $var if it's set $planB if it's not
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param mixed $var        Reference to the main variable
         * @param mixed $planB      What to return if $var isn't available
         * @param bool  $useNullget If set to true, will use self::nullget(), otherwise will use self::get() to
         *                          determinewhether $var is set
         *
         * @return mixed $var if available, $planB if not
         */
        static function ifnull(&$var, $planB, $useNullget = false) {
            $v = $useNullget ? self::nullget($var) : self::get($var);

            return $v !== null ? $v : $planB;
        }

        /**
         * Returns the value of the constant with the name of $const if it's defined, $planB if it's not
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string $const Constant name
         * @param mixed  $planB What to return if $const isn't defined
         *
         * @return mixed
         * @since  1.1
         */
        static function ifundefined($const, $planB) {
            return defined($const) ? constant($const) : $planB;
        }

        /**
         * Checks if the request was made via AJAX
         *
         * @author Art <a.molcanovas@gmail.com>
         * @return bool
         */
        static function isAjaxRequest() {
            return self::get($_SERVER['HTTP_X_REQUESTED_WITH']) == 'XMLHttpRequest';
        }

        /**
         * Returns a hashed browser fingerprint
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string $hashAlgo Hash algorithm to use
         *
         * @return string
         * @since  1.2
         */
        static function getFingerprint($hashAlgo = 'sha256') {
            return hash($hashAlgo,
                        '#QramRAN7*s%6n%@x*53jVVPsnrz@5MY$49o^mhJ8HqY%3a09yJnSWg9lBl$O4CKUb&&S%EgYBjhUZEbhquw$keCjR6I%zMcA!Qr' .
                        self::get($_SERVER['HTTP_USER_AGENT']) .
                        'OE2%fWaZh4jfZPiNXKmHfUw6ok6Z0s#PInaFa8&o#xh#nVyaFaXHPUcv^2y579PnYr5AOs6Zqb!QTAZCgRR968*%QxKc^XNuYYM8' .
                        self::get($_SERVER['HTTP_DNT']) .
                        '%CwyJJ^GAooDl&o0mc%7zbWlD^6tWoNSN&m3cKxWLP8kiBqO!j2PM5wACzyOoa^t7AEJ#FlDT!BMtD$luy%2iZejMVzktaiftpg*' .
                        self::get($_SERVER['HTTP_ACCEPT_LANGUAGE']) .
                        'tep!uTwVXk1CedJq0osEI7p&XCxnC3ipGDWEpTXULEg8J!K1NJSxe4GPor$R3OOb**ZjzPN$$SOHe4ZDcQWQULdtT&XxP2!YYxZy');
        }

        /**
         * Checks if the variable is usable in a foreach loop
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param mixed $input The variable
         *
         * @return bool
         * @since  1.2
         */
        static function isTraversable($input) {
            return is_array($input) || $input instanceof Traversable;
        }

        /**
         * Protects the input from cross-site scripting attacks
         *
         * @author Art <a.molcanovas@gmail.com>
         *
         * @param string|array|Traversable $input The scalar input, or an array/Traversable
         *
         * @return string|array|Traversable The escaped string. If an array or traversable was passed on, the input
         * withall its applicable values escaped.
         * @since  1.3.2 ENT_SUBSTITUTE added<br/>
         *         1.2
         */
        static function unXss($input) {
            if (self::isTraversable($input)) {
                foreach ($input as &$i) {
                    $i = self::unXss($i);
                }
            } elseif (is_scalar($input)) {
                $input = htmlspecialchars($input, ENT_QUOTES | ENT_HTML5 | ENT_SUBSTITUTE);
            }

            return $input;
        }
    }
